windows/code/DBStorage.cpp

```diff
@@ -351,174 +339,236 @@ winrt::Windows::Foundation::IAsyncAction DBStorage::RunTasks()
-            return;
-        }
+        CHECK(BindString(statement, i + 1, keys[i]));
     }
-    for (auto stepResult = sqlite3_step(pStmt.get()); stepResult != SQLITE_DONE;
-         stepResult = sqlite3_step(pStmt.get())) {
+    for (;;) {
+        auto stepResult = sqlite3_step(statement.get());
+        if (stepResult == SQLITE_DONE) {
+            break;
+        }
         if (stepResult != SQLITE_ROW) {
-            InvokeError(m_callback, sqlite3_errmsg(db));
-            return;
+            return m_errorHandler.AddError(sqlite3_errmsg(db));
         }
     }
-    std::vector<winrt::JSValue> callbackParams;
-    callbackParams.push_back(winrt::JSValueArray());
-    m_callback(callbackParams);
+    return true;
 }
 
-void DBStorage::DBTask::getAllKeys(sqlite3 *db)
+std::optional<std::vector<std::string>> DBStorage::DBTask::GetAllKeys(sqlite3 *db) noexcept
 {
-    winrt::JSValueArray result;
-    auto getAllKeysCallback = [&](int cCol, char **rgszColText, char **) {
-        if (cCol >= 1) {
-            result.push_back(rgszColText[0]);
+    CHECK(m_errorHandler.GetErrors().empty());
+    std::vector<std::string> result;
+    auto getAllKeysCallback = [&](int columnCount, char **columnTexts, char **) {
+        if (columnCount > 0) {
+            result.emplace_back(columnTexts[0]);
         }
         return SQLITE_OK;
     };
 
-    if (Exec(db, m_callback, u8"SELECT key FROM AsyncLocalStorage", getAllKeysCallback)) {
-        auto writer = winrt::MakeJSValueTreeWriter();
-        result.WriteTo(writer);
-        std::vector<winrt::JSValue> callbackParams;
-        callbackParams.push_back(nullptr);
-        callbackParams.push_back(winrt::TakeJSValue(writer));
-        m_callback(callbackParams);
-    }
+    CHECK(Exec(db, m_errorHandler, "SELECT key FROM AsyncLocalStorage", getAllKeysCallback));
+    return result;
 }
 
-void DBStorage::DBTask::clear(sqlite3 *db)
+std::optional<bool> DBStorage::DBTask::RemoveAll(sqlite3 *db) noexcept
 {
-    if (Exec(db, m_callback, u8"DELETE FROM AsyncLocalStorage")) {
-        std::vector<winrt::JSValue> callbackParams;
-        callbackParams.push_back(nullptr);
-        m_callback(callbackParams);
+    CHECK(m_errorHandler.GetErrors().empty());
+    CHECK(Exec(db, m_errorHandler, "DELETE FROM AsyncLocalStorage"));
+    return true;
+}
+
+// Read KeyValue from a JSON array.
+void ReadValue(const winrt::IJSValueReader &reader,
+               /*out*/ DBStorage::KeyValue &value) noexcept
+{
+    if (reader.ValueType() == winrt::JSValueType::Array) {
+        int index = 0;
+        while (reader.GetNextArrayItem()) {
+            if (index == 0) {
+                ReadValue(reader, value.Key);
+            } else if (index == 1) {
+                ReadValue(reader, value.Value);
+            } else {
+                // Read the array till the end to keep reader in a good state.
+                winrt::SkipValue<winrt::JSValue>(reader);
+            }
+            ++index;
+        }
     }
```

Comment:
We cannot throw an error because these methods are noexcept and it will cause the app crash. I think we need to add a method to IJSValueReader to be able to report such errors. I have created [RNW #9216](https://github.com/microsoft/react-native-windows/issues/9216) for myself to track it.